src/charm.py

```diff
@@ -7,59 +7,162 @@
+        )
+        self.snap.start_snap_service(snap_service=CHARM_KEY)
+
+        # Active if above commands succeeded, else Maintenance
+        self.unit.status = self.snap.status
+
+        # unit flags itself as 'started' so it can be retrieved by the leader
+        self.cluster.relation.data[self.unit].update(unit_config)
+        self.cluster.relation.data[self.unit].update({"state": "started"})
+
+    def _on_cluster_relation_updated(self, event: EventBase) -> None:
+        """Handler for events triggered by changing units.
+
+        This includes:
+            - Adding ready-to-start units to app data
+            - Updating ZK quorum config
+            - Updating app data state
+        """
+        if not self.unit.is_leader():
+            return
+
+        # avoids the leader removing other units on it's departure
+        # TODO: this is REALLY slow if removing only the quorum leader
+        # the new leader can't find leader until the old unit fully departs
+        # and waits for next update_status event, whenever that may be
+        # try to speed this up, maybe by emitting another event? Reemit?
+        if getattr(event, "departing_unit", None) == self.unit:
+            return
+
+        # units need to exist in the app data to be iterated through for next_turn
+        for unit in self.cluster.started_units:
+            unit_id = self.cluster.get_unit_id(unit)
+            current_value = self.cluster.relation.data[self.app].get(str(unit_id), None)
+
+            # sets to "added" for init quorum leader, if not already exists
+            # may already exist if during the case of a failover of unit 0
+            if unit_id == 0:
+                self.cluster.relation.data[self.app].update(
+                    {str(unit_id): current_value or "added"}
+                )
+
+        # adds + removes members for all self-confirmed started units
+        updated_servers = self.cluster.update_cluster()
+        logger.debug(f"{updated_servers=}")
+
+        # either Active if successful, else Maintenance
+        self.unit.status = self.cluster.status
+
+        if self.cluster.status == ActiveStatus():
+            self.cluster.relation.data[self.app].update(updated_servers)
+        else:
+            # in the event some unit wasn't started/ready
+            event.defer()
+            return
+
+    def _on_get_zookeeper_properties_action(self, event: ActionEvent) -> None:
         """Handler for users to copy currently active config for passing to `juju config`."""
-        msg = self.snap.get_merged_properties(property_label="zookeeper")
+        # TODO: this needs updating to use a client for the config
+        config_map = self.snap.get_properties(property_label="zookeeper")
+        msg = "\n".join([f"{k}={v}" for k, v in config_map.items()])
         event.set_results({"properties": msg})
 
-    def _on_get_snap_apps_action(self, event) -> None:
+    def _on_get_snap_apps_action(self, event: ActionEvent) -> None:
```

Comment:
Just in case a user wants to know what services are exposed from the installed Snap. In the case of Kafka, there's only the two, but if we make changes to the base snap to expose some of the other shell scripts packaged by Apache (e.g ZK Shell, Topics etc) it might be useful as they're not enormously well documented to a user unfamiliar with the Snap. Happy to remove it though as you're right, it's not doing much right now.

Reply:
Yeh, I can't think of a reason why the Juju operator/admin would want this - they should be accessing functionality through actions/config rather than the snap itself?